libparsec/crates/client/src/monitors/workspace_inbound_sync.rs

```diff
@@ -33,104 +37,170 @@ pub(crate) async fn start_workspace_inbound_sync_monitor(
     .await
 }
 
-fn task_future_factory(
+#[derive(Debug)]
+enum IncomingEvent {
+    RemoteChange { entry_id: VlobID },
+    MissedServerEvents,
+}
+
+/// This internal trait is used to abstract all side effets `inbound_sync_monitor_loop` relies on.
+/// This way tests can be done with a mocked version of `WorkspaceOps`&co.
+///
+/// Also note all the methods here are asynchronous (even if some are implemented by wrapping
+/// synchronous call, e.g. `event_bus.send(...)`).
+/// This is needed in the tests given aborting a task is a fire-and-forget call, and hence
+/// a lot (or very few depending on the run) can occur if we don't have a way to detect the
+/// abort has been called and then sleep forever waiting for our task to be actually aborted.
+trait InboundSyncManagerIO: Send + Sync + 'static {
+    async fn wait_for_next_incoming_event(&self) -> Result<IncomingEvent, ()>;
+    async fn retry_later_busy_entry(&mut self, entry_id: VlobID);
+    async fn event_bus_wait_server_online(&self);
+    async fn event_bus_send(&self, event: &impl Broadcastable);
+    fn workspace_ops_refresh_realm_checkpoint(
+        &self,
+    ) -> impl std::future::Future<Output = Result<(), WorkspaceSyncError>> + Send;
+    fn workspace_ops_get_need_inbound_sync(
+        &self,
+        limit: u32,
+    ) -> impl std::future::Future<
+        Output = Result<Vec<VlobID>, WorkspaceGetNeedInboundSyncEntriesError>,
+    > + Send;
+    fn workspace_ops_inbound_sync(
+        &self,
+        entry_id: VlobID,
+    ) -> impl std::future::Future<Output = Result<InboundSyncOutcome, WorkspaceSyncError>> + Send;
+}
+
+/// This is the only implementation of `InboundSyncManagerIO` you should care about !
+struct RealInboundSyncManagerIO {
     workspace_ops: Arc<WorkspaceOps>,
     event_bus: EventBus,
-) -> impl Future<Output = ()> {
-    enum Action {
-        RemoteChange { entry_id: VlobID },
-        MissedServerEvents,
-    }
-    // Channel starts empty, hence the monitor will stay idle until the connection
-    // monitor triggers its initial `EventMissedServerEvents` event
-    let (tx, rx) = channel::unbounded::<Action>();
+    incoming_events_tx: channel::Sender<IncomingEvent>,
+    incoming_events_rx: channel::Receiver<IncomingEvent>,
+    /// Lifetimes are never accessed, but must be kept around for the whole time we
+    /// need to listen to the events.
+    _incoming_events_connection_lifetimes: (
```

Comment:
No because phantom data has no existence at runtime, while here we rely on the lifetime's drop to execute some code

(as sidenote, `EventBusConnectionLifetime` itself uses phatom data to be parametrized by the type of event it is connected to)